### PR TITLE
fixed matrix derivatives and .diff( )

### PR DESCRIPTION
**Pull request by Upabjojr**
- `MatrixExpr` now inherits `Expr`
- Matrix expressions can now be derived the usual way (using `diff` and/or `Derivative`).

### CODE REVIEW

**Comment by asmeurer** [comment]
Looks like there are some test failures. The change itself looks good, once we can figure out the tests.

**Comment by Upabjojr** [comment]
Yeah, I'm at it.

File: sympy/simplify/tests/test_trigsimp.py [hunk 1]
```diff
@@ -351,8 +351,9 @@ def test_issue_2827_trigsimp_methods():
     # all methods should work with Basic expressions even if they
     # aren't Expr
     M = Matrix.eye(1)
+    # NOTE: "groebner" is currently not tested as it raises an exception:
     assert all(trigsimp(M, method=m) == M for m in
-        'fu matching groebner old'.split())
```
Comment (Upabjojr) [inline sympy/simplify/tests/test_trigsimp.py]: @asmeurer had to disable _groebner_, as the matrix is now a non-commutative polynomial, and apparently they are not supported.

Comment (asmeurer) [reply sympy/simplify/tests/test_trigsimp.py]: Shouldn't `trigsimp` on a matrix just apply to the elements of the matrix?

**Pull request merged**